pyvista/examples/_example_loader.py

```diff
@@ -0,0 +1,538 @@
+"""Abstraction layer for downloading, reading, and loading example files.
+
+The classes and methods in this module define an API for working with either
+a single file or multiple files which may be downloaded and/or loaded as an
+example dataset.
+
+Many example datasets have a straightforward input to output mapping:
+    file -> read -> dataset
+
+However, some file formats require multiple input files for reading (e.g.
+separate data and header files):
+    (file1, file1) -> read -> dataset
+
+Or, an example may combine two separate datasets:
+    file1 -> read -> dataset1 ┬─> combined_dataset
+    file2 -> read -> dataset2 ┘
+
+In some cases, the input may be a folder instead of a file (e.g. DICOM):
+    folder -> read -> dataset
+
+In addition, there may be a need to customize the reading function to read
+files with specific options enabled (e.g. set a time value), or perform
+post-read processing to modify the dataset (e.g. set active scalars).
+
+This module aims to serve these use cases and provide a flexible way of
+downloading, reading, and processing files with a generic mapping:
+    file or files or folder -> fully processed dataset(s) in any form
+
+"""
+
+from __future__ import annotations
+
+from abc import abstractmethod
+import functools
+import os
+from typing import (
+    Any,
+    Callable,
+    Optional,
+    Protocol,
+    Sequence,
+    Tuple,
+    TypeVar,
+    Union,
+    runtime_checkable,
+)
+
+import pyvista
+from pyvista.core._typing_core import NumpyArray
+from pyvista.core.utilities.fileio import get_ext
+
+# define TypeVars for two main class definitions used by this module:
+#   1. classes for single file inputs: T -> T
+#   2. classes for multi-file inputs: (T, ...) -> (T, ...)
+_FilePropStrType_co = TypeVar('_FilePropStrType_co', str, Tuple[str, ...], covariant=True)
+_FilePropIntType_co = TypeVar('_FilePropIntType_co', int, Tuple[int, ...], covariant=True)
+
+
+class _FileProps(Protocol[_FilePropStrType_co, _FilePropIntType_co]):
+    @property
+    @abstractmethod
+    def filename(self) -> _FilePropStrType_co:
```

Comment:
I think the name `filename` is a bit misleading as it implies just the name, but the full path is actually returned. Also, the path can be a folder path. This should probably  be renamed to `path`, which is consistent with the property name for the reader class 
https://docs.pyvista.org/version/stable/api/readers/_autosummary/pyvista.basereader.path#pyvista.BaseReader.path

Reply:
+1 to `path`.

Reply:
Great! This change is implemented already in  https://github.com/pyvista/pyvista/pull/5814, if it's ok, we can leave it as `filename` here and address this as part of that PR.